Feature lattice predicates

tonimueller opened the pull request:
Recently, I need to have the monomer ID on the lattice edge for a reaction mechanism. 
For that purpose I wrote the feature FeatureExcludedVolumeIdOnLattice.h which uses a predicate as template parameter. The predicates are specified within the LatticePredicate.h and up to now, there are four functors which can be passed to the feature. 
Of course, the feature has a test :-)

Commit:
Merge branch 'develop' of https://github.com/tonimueller/LeMonADE into develop

Commit:
*Makes it possible to make bonds over periodic boundaries

Commit:
Merge branch 'issue25' into develop

Commit:
* add interface for getEdges()

Commit:
Add !add_bonds and !remove_bonds to commands for reading and writing

Commit:
*MoleculesRead.h
 added class ReadBreaks
*MoleculesWrite.h
 added class WriteAddBonds and WriteRemoveBonds

Commit:
completed the comments

Commit:
*MoleculesWrite.h
 insert typedef for edge_type

Commit:
*MoveAddScMonomer
 added Vector3D.h

Commit:
TestFeatureJumps.cpp:CheckWrite
 add reflected bond vector for consistency

Commit:
TestFeatureMoleculesIO:ExportWrite
 add reflected bonds for consistency

Commit:
TestFeatureMoleculesIO:ExportWrite
 uncomment erase file

Commit:
MoleculesWrite:bug fixes

Commit:
TestFeatureMoeculesIO:WriteReadAdditionalBonds

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
fix for output

Commit:
AnalyzerWriteBfmFile.h: add interface getCommandType() for writing !add_bonds and !remove_bonds

Commit:
MoleculesWrite.h: WriteRemoveBonds and WriteAddBonds can deal with all write types(APPEND,OVERWRITE,NEWFILE)

Commit:
FeatureMoleculesIO.h: hand over write type to write class

Commit:
TestFeatureMoleculesIO.h: added and edit tests for !remove_bonds and !add_bonds

Commit:
AnalyzerWriteBfm: clean up constructor

Commit:
MoleculesWrite: consistent with AnalyzerWriteBfmFile

Commit:
correct the switch in MoleculesWrite

Commit:
clean up

Commit:
New feature which is used to creates bonds across periodic boundaries which
means a connection of monomers with refolded monomers. Such connections are
not part of the normal bond vector set but the refolded bond vectors are.

Commit:
*FeatureBondsetUnsaveCheck.h
  IsValid()-Check in synchronize to allow bonds between refolded monomers

Commit:
*FeatureBondset.h
  again with IsValidStrongCheck() in synchronize

Commit:
Allow access to bondset by declaring it as protected in FeatureBondset.h

Commit:
Merge branch 'issue28' of https://github.com/tonimueller/LeMonADE into issue28

Commit:
add comments for FeatureBondsetUnsaveCheck.h

Commit:
Merge branch 'issue28' into develop
Now, there is a possibility to make a connection between periodic images
by using the FeatureBondsetUnsaveCheck.h in FeatureMoleculesIO.h.

       geändert:       include/LeMonADE/analyzer/AnalyzerWriteBfmFile.h
       geändert:       include/LeMonADE/core/Molecules.h
       geändert:       include/LeMonADE/core/MoleculesRead.h
       geändert:       include/LeMonADE/core/MoleculesWrite.h
       geändert:       include/LeMonADE/feature/FeatureBondset.h
       neue Datei:     include/LeMonADE/feature/FeatureBondsetUnsaveCheck.h
       geändert:       include/LeMonADE/feature/FeatureMoleculesIO.h
       geändert:       include/LeMonADE/updater/moves/MoveAddMonomerSc.h
       geändert:       tests/feature/TestFeatureAttributes.cpp
       geändert:       tests/feature/TestFeatureJumps.cpp
       geändert:       tests/feature/TestFeatureMoleculesIO.cpp

Commit:
UpdaterAbstractCreate::addRing
The function adds a ring of at least 4 monomers around a parent monomer.

Commit:
add WriteReadAdditionalBondsTest.bfm for tests

Commit:
copy *bfm file to the build/tests/ directory

Commit:
complete functions addRing in UpdaterAbstractCreate.h
start to add Test

tonimueller committed:
Add function addRing to UpdaterAbstractCreate.h for adding a ring around a linear chain.
Add test in TestUpdaterAbstractCreate.cpp

tonimueller committed:
merge branches

tonimueller committed:
solve merge conflicts

tonimueller committed:
optimize addRing function for high densities

tonimueller committed:
FeatureExcludedVolumeSc.h

tonimueller committed:
FeatureMoleculesIOUnsaveCheck is used for connections across periodic boundaries

tonimueller committed:
add condition to avoid x-traps

tonimueller committed:
Merge branch 'AddRing' into develop

tonimueller committed:
Merge branches 'develop' and 'develop' of https://github.com/LeMonADE-project/LeMonADE into develop

tonimueller committed:
LatticePredicates.h:
- functors to write different things on the lattice edges

tonimueller committed:
FeatureExcludedVolumeScIdOnLattice.h:
-is a specialised class for FeatureExcludedVolumeSc which uses LatticePredicates
for the values written on the lattice site
-add FeatureAttributes as a required feature, because some predicates make use of it

tonimueller committed:
FeatureExcludedVolumeScIdOnLattice.h:
- correct the inclusion paths
TestFeatureExcludedVolumeScIdOnLattice.cpp:
- test for the single predicates

MartinWenge commented on include/LeMonADE/core/MoleculesRead.h:
```diff
@@ -152,6 +153,79 @@ void ReadBonds<IngredientsType>::execute()
+ **/
+
+template < class IngredientsType >
+void ReadBreaks<IngredientsType>::execute()
```
To be more precise and comparable to ReadAddBonds/WriteRemoveBonds I really prefere to rename these classes
```suggestion
void ReadRemoveBonds<IngredientsType>::execute()
```

MartinWenge commented on include/LeMonADE/core/MoleculesRead.h:
```diff
@@ -152,6 +153,79 @@ void ReadBonds<IngredientsType>::execute()
+ * @tparam IngredientsType Ingredients class storing all system information.
+ **/
+template < class IngredientsType >
+class ReadBreaks: public ReadToDestination < IngredientsType >
```
```suggestion
class ReadRemoveBonds: public ReadToDestination < IngredientsType >
```

MartinWenge commented on include/LeMonADE/core/MoleculesRead.h:
```diff
@@ -152,6 +153,79 @@ void ReadBonds<IngredientsType>::execute()
 
+/***********************************************************************/
+/**
+ * @class ReadBreaks
```
```suggestion
 * @class ReadRemoveBonds
```

MartinWenge commented on include/LeMonADE/core/MoleculesRead.h:
```diff
@@ -152,6 +153,79 @@ void ReadBonds<IngredientsType>::execute()
+{
+ public:
+	//! Empty Constructor, but delegates call to the Feature
+	ReadBreaks(IngredientsType& destination):ReadToDestination< IngredientsType > (destination){}
```
```suggestion
	ReadRemoveBonds(IngredientsType& destination):ReadToDestination< IngredientsType > (destination){}
```

MartinWenge commented on include/LeMonADE/core/MoleculesRead.h:
```diff
@@ -152,6 +153,79 @@ void ReadBonds<IngredientsType>::execute()
+template < class IngredientsType >
+void ReadBreaks<IngredientsType>::execute()
+{
+  std::cout << "ReadBreaks:execute() : updating " <<std::endl;
```
```suggestion
  std::cout << "ReadRemoveBonds:execute() : updating " <<std::endl;
```

MartinWenge commented on include/LeMonADE/core/MoleculesRead.h:
```diff
@@ -152,6 +153,79 @@ void ReadBonds<IngredientsType>::execute()
+    if(stream.fail()){
+      
+    	std::stringstream errormessage;
+      errormessage<<"ReadBreaks<IngredientsType>::execute()\n"
```
```suggestion
      errormessage<<"ReadRemoveBonds<IngredientsType>::execute()\n"
```

MartinWenge commented on include/LeMonADE/core/MoleculesRead.h:
```diff
@@ -152,6 +153,79 @@ void ReadBonds<IngredientsType>::execute()
+    getline(this->getInputStream(),line);
+  }
+  
+  std::cout << "ReadBreaks:execute() : done " <<std::endl;
```
```suggestion
  std::cout << "ReadRemoveBonds:execute() : done " <<std::endl;
```

MartinWenge commented on include/LeMonADE/feature/FeatureMoleculesIO.h:
```diff
@@ -129,6 +130,7 @@ void FeatureMoleculesIO::exportRead(FileImport<IngredientsType>& fileReader)
 	fileReader.registerRead("!number_of_monomers", new  ReadNrOfMonomers< IngredientsType > (destination) );
 	fileReader.registerRead("!bonds", new  ReadBonds< IngredientsType > (destination) );
 	fileReader.registerRead("!add_bonds", new  ReadBonds< IngredientsType > (destination) );
+	fileReader.registerRead("!remove_bonds", new  ReadBreaks< IngredientsType > (destination) );
```
```suggestion
	fileReader.registerRead("!remove_bonds", new  ReadRemoveBonds< IngredientsType > (destination) );
```

MartinWenge commented on include/LeMonADE/feature/FeatureMoleculesIOUnsaveCheck.h:
```diff
@@ -0,0 +1,174 @@
+	fileReader.registerRead("!number_of_monomers", new  ReadNrOfMonomers< IngredientsType > (destination) );
+	fileReader.registerRead("!bonds", new  ReadBonds< IngredientsType > (destination) );
+	fileReader.registerRead("!add_bonds", new  ReadBonds< IngredientsType > (destination) );
+	fileReader.registerRead("!remove_bonds", new  ReadBreaks< IngredientsType > (destination) );
```
```suggestion
	fileReader.registerRead("!remove_bonds", new  ReadRemoveBonds< IngredientsType > (destination) );
```

MartinWenge commented on include/LeMonADE/utility/LatticePredicates.h:
```diff
@@ -0,0 +1,84 @@
+    return LatticeEntry;
+  }
+};
+#endif
```
adding newline to eof
```suggestion
#endif

```

MartinWenge commented on include/LeMonADE/utility/LatticePredicates.h:
```diff
@@ -0,0 +1,84 @@
+ *
+ * @brief Functor returning attribute and monomer id .
+ *
+ * @details Thie functor writes on the first two bits the attribute and on the remaining bits the monomer id.
```
just a typo
```suggestion
 * @details This functor writes on the first two bits the attribute and on the remaining bits the monomer id.
```

MartinWenge commented on include/LeMonADE/updater/UpdaterAbstractCreate.h:
```diff
@@ -252,6 +258,206 @@ bool UpdaterAbstractCreate<IngredientsType>::addMonomerInsideConnectedPair(uint3
+    
+      // check new position (excluded volume, other features)
+      if(addmove.check(ingredients)==true){
+	//check the new bondvector bewten the new monomer and indexB
```
just a typo
```suggestion
	//check the new bondvector between the new monomer and indexB
```

MartinWenge commented on include/LeMonADE/updater/UpdaterAbstractCreate.h:
```diff
@@ -252,6 +258,206 @@ bool UpdaterAbstractCreate<IngredientsType>::addMonomerInsideConnectedPair(uint3
 }
+/******************************************************************************/
+/**
+ * @brief function to add a new monomer two already existing ones.
```
Can you add in this comment, that all monomers are supposed to be connected afterwards

tonimueller replied:
```suggestion
 * @brief function to add a new monomer to two already existing ones such that all three monomers are connected.
```

tonimueller replied:
```suggestion
 * @brief function to add a new monomer to two already existing ones such that all three are connected afterwards
```

MartinWenge reviewed (changes requested):
Nice feature, I really like the additions to UpdaterAbstractCreate :+1: 

Please check the renaming of the readBreaks class, I would prefere it.

I passed all the tests and my own tool work with this version. Go for it :o)

Commit:
Update include/LeMonADE/core/MoleculesRead.h

Co-Authored-By: tonimueller <tm.toni.mueller@googlemail.com>

Commit:
Apply suggestions from code review

Use a consistent class naming for RemovedBonds and AddBonds and correct some typos.

Co-Authored-By: tonimueller <tm.toni.mueller@googlemail.com>

tonimueller reviewed (commented):
Changed the comment for the function 'addMonomerInsideConnectedPair'

Commit:
Update include/LeMonADE/updater/UpdaterAbstractCreate.h

tonimueller commented:
I accepted all changes and the pull request can be closed I think
